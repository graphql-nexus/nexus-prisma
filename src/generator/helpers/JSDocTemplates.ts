import { DMMF } from '@prisma/client/runtime'
import endent from 'endent'
import { repeat } from 'lodash'

export function jsDocForModel(model: DMMF.Model): string {
  return model.documentation
    ? endent`
        /**
         * ${model.documentation}
         */
      `
    : missingModelDocumentation(model)
}

export function jsDocForField({ field, model }: { field: DMMF.Field; model: DMMF.Model }): string {
  return field.documentation
    ? endent`
        /**
         * ${field.documentation}
         */
      `
    : missingFieldDocumentation({ field, model })
}

export function missingModelDocumentation({ name }: DMMF.Model): string {
  return endent`
    /**
     * ### ðŸ“” Missing Model Documentation for \`${name}\`
     * 
     * Get JSDoc documentation for this model automatically by documenting
     * it in your Prisma Schema âœ¨!
     * 
     * To document a model add a tripple slash comment above it.
     * 
     * Learn more about Prisma Schema comments [here](https://www.prisma.io/docs/concepts/components/prisma-schema#comments).
     * 
     * ${jsDocDiviion('Example 1')}
     * 
     * \`\`\`prisma
     * /// Lorem ipsum dolor sit amet...
     * model ${name} {
     *   id  String  @id
     * }
     * \`\`\`
     * 
     * ${jsDocDiviion('Example 2')}
     * 
     * \`\`\`prisma
     * /// A user record. A user maps 1:1 with a person. Not
     * /// to be confused with an \`Account\` which on person
     * /// may have multiple of, all linked to a single \`User\`.
     * model User {
     *   id  String  @id
     * }
     * \`\`\`
     */
  `
}

export function missingFieldDocumentation({
  field,
  model,
}: {
  field: DMMF.Field
  model: DMMF.Model
}): string {
  return endent`
    /**
     * ### ðŸ“” Missing Field Documentation for \`${field.name}\`
     * 
     * Get JSDoc documentation for this field automatically by documenting
     * it in your Prisma Schema âœ¨!
     * 
     * To document a field add a tripple slash comment above it.
     * 
     * Learn more about Prisma Schema comments [here](https://www.prisma.io/docs/concepts/components/prisma-schema#comments).
     * 
     * ${jsDocDiviion('Example 1')}
     * 
     * \`\`\`prisma
     * model ${model.name} {
     *   /// Lorem ipsum dolor sit amet.
     *   ${field.name}  ${field.type}${field.isRequired ? '' : '?'}
     * }
     * \`\`\`
     * 
     * ${jsDocDiviion('Example 2')}
     * 
     * \`\`\`prisma
     * model User {
     *   /// Identifier for a user. Will never change once set.
     *   /// More stable than the \`handle\` field  which user
     *   /// _can_ change, but is not human-friendly. Prefer this
     *   /// for machine consumers but prefer \`handle\` for human
     *   /// visible things like URL slugs.
     *   id  String  @id
     * }
     * \`\`\`
     */
  `
}

/**
 * Render a centered title with bars on either side filling left over space.
 *
 * Sized according to standard JSDoc window size in VSCode.
 */
function jsDocDiviion(title: string): string {
  const titlePadding = ' '
  const barLength = Math.floor(MAX_JS_DOC_LENGTH / 2 - title.length - titlePadding.length * 2)
  const bar = repeat('â€“', barLength)
  return endent`
    #### ${bar}${titlePadding}${title}${titlePadding}${bar}
  `
}

const MAX_JS_DOC_LENGTH = 72
