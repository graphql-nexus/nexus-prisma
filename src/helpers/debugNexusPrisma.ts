import debug from 'debug'

const debugName = 'nexus-prisma'

export const d = debug(debugName)

d.log = console.log.bind(console)

export function debugNexusPrisma(...names: string[]) {
  const d2 = debug([debugName].concat(names).join(':'))
  d2.log = console.log.bind(console)
  return d2
}

export function packageRun() {}
