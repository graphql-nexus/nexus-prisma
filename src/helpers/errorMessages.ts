export const GITHUB_NEW_ISSUE_LINK = `https://github.com/prisma/nexus-prisma/issues/new/choose`

export const GITHUB_NEW_DISCUSSION_LINK = `https://github.com/prisma/nexus-prisma/discussions/new`
