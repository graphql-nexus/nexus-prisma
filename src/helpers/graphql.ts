export type GraphQLScalarType = 'ID' | 'String' | 'Int' | 'Float' | 'Boolean'

export const graphQLScalarTypes = {
  ID: 'ID',
  String: 'String',
  Float: 'Float',
  Boolean: 'Boolean',
  Int: 'Int',
} as const
