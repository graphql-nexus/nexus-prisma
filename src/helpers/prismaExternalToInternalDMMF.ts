// @ts-expect-error Private api
import * as PrismaClientGenerator from '@prisma/client/generator-build'
import { DMMF, DMMF as ExternalDMMF } from '@prisma/generator-helper'

export const externalToInternalDmmf = PrismaClientGenerator.externalToInternalDmmf as (
  document: ExternalDMMF.Document
) => DMMF.Document
