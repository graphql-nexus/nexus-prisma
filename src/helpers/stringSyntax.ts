import endent from 'endent'

export function quote(x: string): string {
  return `'${x}'`
}

export function apply(functionName: string, parameters: string | string[]): string {
  return endent`
    ${functionName}(${arrayify(parameters).join(', ')})
  `
}

function arrayify<T>(x: T): T extends any[] ? T : T[] {
  if (Array.isArray(x)) {
    return x as any
  }
  return [x] as any
}
