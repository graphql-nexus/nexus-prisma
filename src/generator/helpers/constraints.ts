import { DMMF } from '@prisma/client/runtime'
import { RecordUnknown } from '../../helpers/utils'

/**
 * Find the unique identifiers necessary to indentify a field
 *
 * Unique fields for a model can be one of (in this order):
 * 1. One (and only one) field with an @id annotation
 * 2. Multiple fields with @@id clause
 * 3. One (and only one) field with a @unique annotation (if there are multiple, use the first one)
 * 4. Multiple fields with a @@unique clause
 */
export function resolveUniqueIdentifiers(model: DMMF.Model): string[] {
  // Try finding 1.
  const singleIdField = model.fields.find((f) => f.isId)

  if (singleIdField) {
    return [singleIdField.name]
  }

  // Try finding 2.
  if (model.idFields && model.idFields.length > 0) {
    return model.idFields
  }

  // Try finding 3.
  const singleUniqueField = model.fields.find((f) => f.isUnique)

  if (singleUniqueField) {
    return [singleUniqueField.name]
  }

  // Try finding 4.
  if (model.uniqueFields && model.uniqueFields.length > 0) {
    return model.uniqueFields[0] as string[] // I don't know why typescript want a cast here
  }

  throw new Error(`Unable to resolve a unique identifier for the Prisma model: ${model.name}`)
}

export function findMissingUniqueIdentifiers(
  data: RecordUnknown,
  uniqueIdentifiers: string[]
): string[] | null {
  const missingIdentifiers: string[] = []

  for (const identifier of uniqueIdentifiers) {
    if (!data[identifier]) {
      missingIdentifiers.push(identifier)
    }
  }

  if (missingIdentifiers.length > 0) {
    return missingIdentifiers
  }

  return null
}

export function buildWhereUniqueInput(data: RecordUnknown, uniqueIdentifiers: string[]): RecordUnknown {
  if (uniqueIdentifiers.length === 1) {
    return pickFromRecord(data, uniqueIdentifiers)
  }

  const compoundName = uniqueIdentifiers.join('_')

  return {
    [compoundName]: pickFromRecord(data, uniqueIdentifiers),
  }
}

function pickFromRecord(record: RecordUnknown, keys: string[]) {
  const output: Record<string, unknown> = {}

  for (const identifier of keys) {
    if (record[identifier]) {
      output[identifier] = record[identifier]
    }
  }

  return output
}
