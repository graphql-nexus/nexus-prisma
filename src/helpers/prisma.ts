import { DMMF } from '@prisma/client/runtime'
import dedent from 'dindist'
import kleur from 'kleur'
import ono from 'ono'
import { inspect } from 'util'
import { detectProjectPackageManager, renderRunBin } from '../lib/packageManager'
import { d } from './debugNexusPrisma'
import { GITHUB_NEW_DISCUSSION_LINK } from './errorMessages'

/**
 * Given a package loader, attempt to get the Prisma Client DMMF.
 *
 * @remarks Only the given require function is truly import, the rest is used for better error messages.
 *
 *          This function is designed to support working with bundlers. Specifically `ncc` has been
 *          tested.
 *
 *          This function intentionally does not do the `require`/`import` itself, leaving that to
 *          upstream code to handle in static ways that bundlers will be able to process.
 */
export const getPrismaClientDmmf = (packageLoader: {
  /**
   * A function that must return the Prisma Client Package
   */
  require: () => unknown
  /**
   * The import specifier being used (the from "..." part)
   */
  importId: string
  /**
   * The resolved import specifier being used. This can be different than important ID in two ways:
   *
   * 1. NodeJS lookp algorithm
   * 2. Bundlers that rewrite import paths
   */
  importIdResolved: string
}): DMMF.Document => {
  d('get dmmf from @prisma/client')

  let prismaClientPackage: unknown

  // prettier-ignore
  const printedImportId = `${kleur.yellow(packageLoader.importId)} (resolved as ${packageLoader.importIdResolved})`

  try {
    // eslint-disable-next-line
    prismaClientPackage = packageLoader.require()
  } catch (error) {
    // prettier-ignore
    throw ono(error as Error, dedent`
      Failed to get Prisma Client DMMF. An error occured while trying to import it from ${printedImportId}.
    `)
  }

  if (!(typeof prismaClientPackage === 'object' && prismaClientPackage !== null)) {
    // prettier-ignore
    throw new Error(dedent`
      Failed to get Prisma Client DMMF. It was imported from ${printedImportId} but was not the expected type. Got:

      ${inspect(prismaClientPackage)}
    `)
  }

  const prismaClientPackageObject = prismaClientPackage as Record<string, unknown>

  // eslint-disable-next-line
  if (!prismaClientPackageObject.dmmf) {
    // prettier-ignore
    throw new Error(dedent`
      Failed to get Prisma Client DMMF. It was imported from ${printedImportId} but did not contain "dmmf" data. Got:

      ${inspect(prismaClientPackage)}

      This usually means that you need to run Prisma Client generation. Please run ${renderRunBin(detectProjectPackageManager(), `prisma generate`)}.
      If that does not solve your problem, you can get community help by opening a discussion at ${kleur.yellow(GITHUB_NEW_DISCUSSION_LINK)}.
    `)
  }

  // Simple duck type to sanity check we got good data at runtime.

  const dmmf = prismaClientPackageObject.dmmf as DMMF.Document

  const expectedFields = ['datamodel', 'schema', 'mappings'] as const

  if (expectedFields.find((fieldName) => dmmf[fieldName] && typeof dmmf[fieldName] !== 'object')) {
    throw new Error(dedent`
      The DMMF imported from ${packageLoader.importId} appears to be invalid. Missing one/all of expected fields: 
    `)
  }

  return dmmf
}

export type PrismaScalarType =
  | 'String'
  | 'Boolean'
  | 'Int'
  | 'BigInt'
  | 'Float'
  | 'Decimal'
  | 'DateTime'
  | 'Json'
  | 'Bytes'
