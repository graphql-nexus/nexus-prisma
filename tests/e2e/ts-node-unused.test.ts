import * as Path from 'path'
import dindist from 'dindist'
import { konn, providers } from 'konn'
import 'ts-replace-all'
import { project } from '../__providers__/project'
import { createPrismaSchema } from '../__helpers__/helpers'
import { bindRunOrThrow } from './run-or-throw'

const ctx = konn()
  .useBeforeAll(providers.dir())
  .useBeforeAll(providers.run())
  .useBeforeAll(project())
  .beforeAll(async (ctx) => {
    ctx.fixture.use(Path.join(__dirname, 'fixtures/ts-node-unused'))
    ctx.fs.write(
      `prisma/schema.prisma`,
      createPrismaSchema({
        content: dindist`
          model Foo {
            id  String  @id
          }
        `,
      })
    )
    bindRunOrThrow(ctx)
    ctx.runOrThrow(`${Path.join(process.cwd(), 'node_modules/.bin/yalc')} add ${ctx.thisPackageName}`)
    await ctx.runAsync(`yarn install --legacy-peer-deps`, { env: { PEER_DEPENDENCY_CHECK: 'false' } })
    return ctx
  })
  .done()

it('when project does not have ts-node installed nexus-prisma generator still generates if there are no TS generator config files present', async () => {
  const result = await ctx.runOrThrowPackageScript(`build`)
  expect(normalizeGeneratorOutput(result.stdout)).toMatchSnapshot()
})

const normalizeGeneratorOutput = (output: string) =>
  output
    .replaceAll(/(\d+|\d+(\.\d+))(ms|s)/g, '<SOME TIME><unit>')
    .replaceAll(/ to .* in /g, ' to <SOME PATH> in ')
    .replaceAll(/loaded from.*/g, 'loaded from <SOME PATH>')
    .replaceAll(/Generated Prisma Client \(.*\)/g, 'Generated Prisma Client (<SOME VERSION>)')
