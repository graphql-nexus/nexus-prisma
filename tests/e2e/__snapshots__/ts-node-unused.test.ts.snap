// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when project does not have ts-node installed nexus-prisma generator still generates if there are no TS generator config files present 1`] = `
"Prisma schema loaded from <SOME PATH>
You can now start using Nexus Prisma in your code. Reference: https://pris.ly/d/nexus-prisma

✔ Generated Prisma Client (<SOME VERSION>) to <SOME PATH> in <SOME TIME>ms

✔ Generated Nexus Prisma to <SOME PATH> in <SOME TIME>ms
You can now start using Prisma Client in your code. Reference: https://pris.ly/d/client
\`\`\`
import { PrismaClient } from '@prisma/client'
const prisma = new PrismaClient()
\`\`\`"
`;
